### PR TITLE
[4.2] Remove the SCSS deprecations

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes

- Removing the math.div deprecations from the scss compiler 
- update Font Awesome to v6
- update choices.js to v10

### Testing Instructions
- Apply this Pr's branch
- run `npm ci`
- Check that nothing is broken visually

### CODE REVIEW

**Comment by HarshJaiswani** [comment]
I have tested this item :white_check_mark: successfully on bcc3ce8fb6197b16c1768f139c60842d07c2f49c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36906">issues.joomla.org/tracker/joomla-cms/36906</a>.</sub>

File: package-lock.json [hunk 1]
```diff
@@ -10463,9 +10410,9 @@
       }
     },
     "codemirror": {
-      "version": "5.65.1",
-      "resolved": "https://registry.npmjs.org/codemirror/-/codemirror-5.65.1.tgz",
-      "integrity": "sha512-s6aac+DD+4O2u1aBmdxhB7yz2XU7tG3snOyQ05Kxifahz7hoxnfxIRHxiCSEv3TUC38dIVH8G+lZH9UWSfGQxA=="
+      "version": "5.64.0",
```
Comment (brianteeman) [inline package-lock.json]: this should be 5.65.1

Comment (dgrammatiko) [reply package-lock.json]: done, btw not reflecting the actual version in the package.json is the source of those inconsistencies...

**Comment by HLeithner** [comment]
2 questions why do you use * .5 instead of / 2 ? And isn't awesome 6.0 a b/c break. Iirc fontawesome has some really strange naming breaks for some icons in major versions I might be wrong

**Comment by brianteeman** [comment]
> 2 questions why do you use * .5 instead of / 2 ? 

[Because slash is used as both a separator and as division](https://sass-lang.com/documentation/breaking-changes/slash-div)

**Comment by dgrammatiko** [comment]
About FA6: their changelog doesn't mention any major breaks: https://fontawesome.com/docs/changelog/

EDIT: FA6 changed the names of some icons, the list is [here](https://fontawesome.com/docs/web/setup/upgrade/whats-changed#icons-renamed-in-version-6) but they provide a [scss file](https://github.com/FortAwesome/Font-Awesome/blob/6.x/scss/_shims.scss) that bridges the names (use both names, v5 + v6). I will update this PR to include this later today 

About choices.js (which also has a major version update), their changelog also doesn't state any major B/C breaks: https://github.com/Choices-js/Choices/releases/tag/v10.0.0

About the slash on the scss @brianteeman already posted their post for the why, The TL:DR; is  the `/` is a CSS modifier but in SCSS was also used for math devision. The math division now should be possible only using `math.div($x, $y);`...

File: build/media_source/com_media/scss/components/_media-infobar.scss [hunk 2]
```diff
@@ -27,16 +27,16 @@ $fa-css-prefix:    fa;
...
   }
   dt, dd {
     position: relative;
     direction: ltr;
     width: 100%;
     min-height: 1px;
-    padding-right: ($col-gutter-width / 2);
-    padding-left: ($col-gutter-width / 2);
+    padding-right: ($col-gutter-width * .5);
+    padding-left: ($col-gutter-width * .5);
```
Comment (Quy) [inline build/media_source/com_media/scss/components/_media-infobar.scss]:
```suggestion
    padding-right: $col-gutter-width * .5;
    padding-left: $col-gutter-width * .5;
```

File: build/media_source/templates/site/cassiopeia/scss/vendor/_chosen.scss [hunk 3]
```diff
@@ -97,7 +97,7 @@ $chosen-select-padding-y: $form-select-padding-y + .21;
         &::before {
           position: absolute;
           top: 6px;
-          right: ($chosen-multi-close-width / 4);
+          right: ($chosen-multi-close-width * .25);
```
Comment (Quy) [inline build/media_source/templates/site/cassiopeia/scss/vendor/_chosen.scss]:
```suggestion
          right: $chosen-multi-close-width * .25;
```

File: build/media_source/templates/site/cassiopeia/scss/vendor/metismenu/_metismenu.scss [hunk 4]
```diff
@@ -138,8 +138,8 @@
             position: relative;
             top: 0;
             flex-basis: calc(100% + #{$cassiopeia-grid-gutter});
-            margin-top: ($cassiopeia-grid-gutter/2);
-            margin-bottom: ($cassiopeia-grid-gutter/-2);
+            margin-top: ($cassiopeia-grid-gutter * .5);
+            margin-bottom: ($cassiopeia-grid-gutter * -.5);
```
Comment (Quy) [inline build/media_source/templates/site/cassiopeia/scss/vendor/metismenu/_metismenu.scss]:
```suggestion
            margin-top: $cassiopeia-grid-gutter * .5;
            margin-bottom: $cassiopeia-grid-gutter * -.5;
```

**Comment by dgrammatiko** [comment]
@HLeithner the `shims` file is imported so this should be 100% B/C

**Comment by Quy** [comment]
Please fix conflicts.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 27eaa7b23e12684c379ba02546c3ba1063b15ce7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36906">issues.joomla.org/tracker/joomla-cms/36906</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 27eaa7b23e12684c379ba02546c3ba1063b15ce7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36906">issues.joomla.org/tracker/joomla-cms/36906</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36906">issues.joomla.org/tracker/joomla-cms/36906</a>.</sub>

**Comment by brianteeman** [comment]
> Check that nothing is broken visually

### Before 1
image

### After 1
image

### Before 2
image

### After 2
image

### Before 3 - weight?
image


### After 3 - weight?
image

There are LOTS of other icon changes. In general they probably dont matter but where we have maybe used an icon for something other than its original purpose the changes are not good.

_I thought they said there was a b/c shim in fontawesom6_

**Comment by dgrammatiko** [comment]
> I thought they said there was a b/c shim in fontawesom6

There is and we are using it but it covers the name changes. For the visual changes someone needs to diff both icon sets and do any adjustments, if there is a new name that fits better the existing mapping. I did that for icomoon and quite frankly I'm not up to the task this time as it takes forever (5000+ icons). I'm not saying that it shouldn't happen but I don't have the time for this atm, so if this is a deal breaker someone else needs to take over here

**Comment by brianteeman** [comment]
>  so if this is a deal breaker someone else needs to take over here

I am just the messenger ;)

**Comment by bembelimen** [comment]
I like the update, but will not include a new FA version in a patch release, so will rebase to 4.2 and let @roland-d and @fancyFranci decide how the changed icons should be handled.

**Comment by roland-d** [comment]
@dgrammatiko I sent you a PR to fix the conflict, after that I can merge this PR. Thank you.

**Comment by dgrammatiko** [comment]
>  after that I can merge this PR

About merging this PR: @brianteeman already spotted some visual changes due to the upgrade to v6 of Font Awesome. This might need some research (it's just comparing the font icon names using the 2 different sites v5 and v6, making some notes and then maybe figure out if some other mapping is making more sense for the current use case, it's a bit of work...)

It would be nice to roll an article that 4.2 will use FA v6 and ask devs to adjust (early) their code eg from:
```css
font-family: "Font Awesome 5 Free";
```

to 

```css
font-family: "Font Awesome 6 Free","Font Awesome 5 Free";
```

Also this is the only part that needs to be documented

**Comment by Quy** [comment]
Please merge. I noticed it too, however, this can be done in separate PR which I can do.

**Comment by roland-d** [comment]
Thank you @Quy I am just waiting for Drone and AppVeyor to finish.

**Comment by dgrammatiko** [comment]
@roland-d do you want me to do some back port of this to 4.1 (without the FA v6) or someone already on it?

**Comment by brianteeman** [comment]
do/should we really be updating fontawesome? \they are breaking changes AND fs5 is designated LTS

**Comment by dgrammatiko** [comment]
> do/should we really be updating fontawesome? \they are breaking changes AND fs5 is designated LTS

FA v5 still has the deprecated div notifications and there is not update but also it's not the end of the world we could override some values locally and fix it in this repo if there's no fix upstream. Anyways, not my decision, I could redo this with the v5...

**Comment by brianteeman** [comment]
>  the deprecated div notifications
which are not seen by the user

**Comment by dgrammatiko** [comment]
> which are not seen by the user

True, but it's annoying for devs (not really a reason) and might be a problem for new contributors (not really, if there was a nice welcoming DX)
Also there's no documentation that states that the node audit/deprecations/etc are irrelevant for the distributed product

FWIW this PR's intention was to just fix the math div deprecations, the FA upgrade shouldn't be part of this. Probably I was too fast to jump on the newest-greatest train without thinking too much. I'll try to do a PR in the 4.1 with only the needed changes and close this. If there's a decision to go with v6, diffing the changes here and making a new pr should be a 5min job...

**Comment by brianteeman** [comment]
agreed

**Comment by brianteeman** [comment]
and 6,1 comes out next week

**Comment by dgrammatiko** [comment]
OK closing here the PR for 4.1 is here: https://github.com/joomla/joomla-cms/pull/37243

If anyone still wants to upgrade to FA v6 the code is here

**Comment by joomleb** [comment]
Hi guys, 
sorry,I did not understand. At the end FA6 will not be included into the J4.1, right ?
But, Will it be added in J4.2 ?

**Comment by roland-d** [comment]
Hello,

This will not be in Joomla 4.2 nor do I know when it will come.

**Comment by HDInfautre** [comment]
Hi
There are lots of new icons :-)
Do you know when this can be integrated?
4.2.3 ? or what is the method to do it ? if there is no date.

**Comment by HLeithner** [comment]
Font Awesome icon packs are not backward compatible, so upgrading them would break other websites. if we upgrade it, it's 
likely to be in Joomla 5.0.

**Comment by HDInfautre** [comment]
> Font Awesome icon packs are not backward compatible, so upgrading them would break other websites. if we upgrade it, it's likely to be in Joomla 5.0.
Hi
Sgniff, thank you for the clarification.

**Pull request closed without merging**